File: 2019/cpp/day_06a.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

void recursiveOrbitCount(
    const std::string& current_object,
    const std::unordered_map<std::string, std::unordered_set<std::string>>&
        is_orbitted_by,
    std::unordered_map<std::string, int>& n_orbits) {
  if (is_orbitted_by.find(current_object) == std::end(is_orbitted_by)) {
    n_orbits[current_object] = 0;
  } else {
    n_orbits[current_object] = 0;
    const auto ele_it = is_orbitted_by.find(current_object);
    for (const auto orbitting_object : ele_it->second) {
      recursiveOrbitCount(orbitting_object, is_orbitted_by, n_orbits);
      n_orbits[current_object] += n_orbits[orbitting_object] + 1;
    }
  }
}

int main(int argc, char* argv[]) {
  // Get input
  std::string input = "../input/day_06_input";
  if (argc > 1) {
    input = argv[1];
  }
  std::ifstream file(input);
  std::unordered_map<std::string, std::string> orbits;
  std::unordered_map<std::string, std::unordered_set<std::string>>
      is_orbitted_by;
  std::unordered_map<std::string, int> n_orbits;

  // Solve
  std::string line;
  const std::string delim = ")";
  while (std::getline(file, line)) {
    const size_t delim_pos = line.find(delim);
    const std::string primary_object = line.substr(0, delim_pos);
    const std::string orbitting_object = line.substr(
        delim_pos + delim.size(), line.size() - delim_pos - delim.size());
    orbits[orbitting_object] = primary_object;
    if (is_orbitted_by.find(primary_object) == std::end(is_orbitted_by)) {
      is_orbitted_by.insert({primary_object, {}});
    }
    is_orbitted_by[primary_object].insert(orbitting_object);
  }

  recursiveOrbitCount("COM", is_orbitted_by, n_orbits);
  size_t total = 0;
  for (const auto& [key, val] : n_orbits) {
    total += val;
  }
  std::cout << total << '\n';
  return total;
}

File: 2019/python/day_06a.py
#!/usr/bin/python3

import sys


def recursiveOrbitCount(current_object, is_orbitted_by, n_orbits):
    if current_object not in is_orbitted_by.keys():
        n_orbits[current_object] = 0
    else:
        n_orbits[current_object] = 0
        for orbitting_object in is_orbitted_by[current_object]:
            recursiveOrbitCount(orbitting_object, is_orbitted_by, n_orbits)
            n_orbits[current_object] += n_orbits[orbitting_object] + 1


def main():
    # Get input
    file_name = "../input/day_06_input"
    if len(sys.argv) > 1:
        file_name = sys.argv[1]
    file = open(file_name)
    orbits = dict()
    is_orbitted_by = dict()
    n_orbits = dict()
    input = list()
    for line in file:
        line = line.strip("\n")
        input.append(line)

    # Solve
    for line in input:
        orbit = line.split(")")
        orbits[orbit[1]] = orbit[0]
        if orbit[0] not in is_orbitted_by.keys():
            is_orbitted_by[orbit[0]] = set()
        is_orbitted_by[orbit[0]].add(orbit[1])
    recursiveOrbitCount("COM", is_orbitted_by, n_orbits)
    total = 0
    for n in n_orbits.values():
        total += n
    print(total)
    return total


if __name__ == "__main__":
    main()
